File: 51-n-queens/51-n-queens.py
class Solution:
   
    def solveNQueens(self, n: int) -> List[List[str]]:
        ans=[[]]
        col=set()
        d1=set()
        d2=set()
        v=[]
        def solve(r):
            if(r==n):
                ans.append(list(v))
                return
            for c in range(0,n):
                if (c not in col and r-c not in d1 and r+c not in d2):
                    col.add(c)
                    d1.add(r-c)
                    d2.add(r+c)
                    s = '.'*n
                    s=list(s);
                    s[c]='Q'
                    v.append("".join(s))
                    solve(r+1)
                    v.pop()
                    col.remove(c)
                    d1.remove(r-c)
                    d2.remove(r+c)
        ans.pop()
        solve(0)
        return ans;

File: 51-n-queens/51-n-queens.cpp
class Solution {
public:
    vector<string>v;
    vector<vector<string>>ans;
    int visCol[10];
    int d1[20];
    int d2[20];
    void solve(int r,int n){
        if(r==n){
            ans.push_back(v);
            return;
        }
        for(int c=0;c<n;c++){
            if(!visCol[c] && !d1[r+c] && !d2[r-c+9]){
                visCol[c] = d1[r+c] = d2[r-c+9] = 1;
                string s=string(n,'.'); s[c]='Q';
                v.push_back(s);
                solve(r+1,n);
                visCol[c] = d1[r+c] = d2[r-c+9] = 0;
                v.pop_back();
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
        memset(visCol,0,sizeof visCol);
        memset(d1,0,sizeof d1);
        memset(d2,0,sizeof d2);
        v.clear();
        ans.clear();
        solve(0,n);
        return ans;
    }
};